move BLAS to a separate backend

slaren opened the pull request:
Moves BLAS support from `ggml.c` to a separate backend, and adds the necessary changes to ggml-backend to support backends that only implement matrix multiplication.

- Changes to ggml-backend
    - Support for fallback to CPU with `ggml_backend_sched`
        - Operations not implemented by a backend will be automatically run on the CPU, as long as the operation is reported as not supported in the `supports_op` function of the backend
    - Moved buffer type function `supports_backend` to backend function `supports_buft`
    - Backends that want to declare compatibility with any kind of host buffer can return `ggml_backend_buft_is_host` from `supports_buft`
    - `ggml_backend_sched` will avoid copies between backends when the backend supports the buffer type
        - Eg. when switching from Metal to CPU, no tensors will be copied since the Metal buffers are compatible with the CPU backend (but not the other way around)
- The `GGML_SCHED_DEBUG` environment variable can be used to view the graph splits. This is useful to see what operations are being run on each backend
- Adds the BLAS backend
    - Supports matrix multiplication using a BLAS library. Previously, this was supported as part of the CPU backend
    - Threads are no longer spinning while BLAS is running, potentially improving performance, and batch processing is no longer limited to 4 threads when using BLAS
    - The number of threads of the BLAS library configured automatically for OpenBLAS and BLIS (with `-t` or `-tb`)
    - For better performance, it is recommended to use OpenMP versions of the BLAS libraries, if available (except macOS)
    - Like before, to enable the BLAS backend, build with the flag `LLAMA_BLAS` when using cmake, or when using make, `LLAMA_OPENBLAS`, `LLAMA_OPENBLAS64` or `LLAMA_BLIS`
    - On macOS, this is enabled by default through Accelerate
    - BLAS support has been removed from the CPU backend in `ggml.c`. Applications that want to support BLAS will need to use the BLAS backend
    - Since this backend only implements matrix multiplication, it should be used with `ggml_backend_sched` alongside the CPU or other backends to provide support for other operations
    - Note: the BLAS backend should not be used alongside GPU backends, as it will prevent offloading of large batches with partial offloading

slaren committed:
move BLAS to a separate backend

ggerganov commented:
> this will also have the effect that using BLAS will require using ggml-backend and ggml_backend_sched, is that a problem?

Will just need to adapt `whisper.cpp` when it's ready

zhouwg commented:
> > this will also have the effect that using BLAS will require using ggml-backend and ggml_backend_sched, is that a problem?
> 
> Will just need to adapt `whisper.cpp` when it's ready

I had been submitted a PR to address this and it works very well with whisper.cpp and llama.cpp using QNN backend.

slaren commented:
@mofosyne I appreciate that you are trying to help, but please don't do that on my PRs. I very often have uncommitted local changes and I prefer to deal with the merge conflicts myself.

slaren committed:
rename GGML_USE_OPENBLAS to GGML_USE_BLAS

slaren committed:
add support for out_prod

slaren committed:
alloc : reuse same buffer when the same buffer type if used multiple times

slaren commented:
@ggerganov I am thinking about how accelerate should interact with the BLAS backend. I think this would make sense:
- `GGML_USE_ACCELERATE` defined: accelerate is used in `ggml.c`
- `GGML_USE_ACCELERATE` and `GGML_USE_BLAS` defined: the BLAS backend is used using accelerate as the BLAS library

Conversely:
- If `GGML_USE_BLAS` is not defined, accelerate will not be used for GEMMs

Currently `llama.cpp` has to check check for `defined(GGML_USE_BLAS) || defined(GGML_USE_ACCELERATE)` to decide when to use the BLAS backend, which doesn't seem very good. In the BLAS backend, accelerate is treated in the same way as any other BLAS library.

slaren committed:
support multithreaded dequantization with std::async when openmp is not available

ggerganov commented:
Yes, that makes sense. With only `GGML_USE_ACCELERATE` we will still be able to use some non-GEMM functionality from the Accelerate framework such as `vDSP` - it's just very convenient for Apple Silicon devices to have this framework available in the core `ggml.c`. For GEMMs we would explicitly need to have `GGML_USE_BLAS` for all kinds of BLAS implementations, including Accelerate's BLAS

slaren committed:
fix apple build

slaren committed:
ggml-ci

slaren committed:
reuse main thread

slaren commented:
Some notes on performance.

BLIS responds very well to these changes, and improves performance in most cases. In master, the number of threads is limited to 4 when using BLAS. This is not the case with the BLAS backend. Using a higher number of threads should improve the performance of dequantizing and other operations.


`LLAMA_BLIS=1 BLIS_NUM_THREADS=24`:

| CPU       | Model            |   Model Size [GiB] |   Threads | Test   |   t/s master |   t/s sl/blas-backend |   Speedup |
|:----------|:-----------------|-------------------:|----------:|:-------|-------------:|----------------------:|----------:|
| i9-13900K | llama 7B Q4_0    |               3.56 |         1 | pp32   |        10.03 |                 11.13 |      1.11 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         1 | pp64   |        17.65 |                 19.41 |      1.10 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         1 | pp128  |        29.66 |                 28.09 |      0.95 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         4 | pp32   |        15.58 |                 20.81 |      1.34 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         4 | pp64   |        26.79 |                 32.48 |      1.21 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         4 | pp128  |        39.35 |                 46.65 |      1.19 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         8 | pp32   |        15.79 |                 22.86 |      1.45 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         8 | pp64   |        27.02 |                 35.84 |      1.33 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         8 | pp128  |        39.06 |                 48.96 |      1.25 |
| i9-13900K | llama 7B Q4_0    |               3.56 |        16 | pp32   |        15.37 |                 22.30 |      1.45 |
| i9-13900K | llama 7B Q4_0    |               3.56 |        16 | pp64   |        26.78 |                 28.70 |      1.07 |
| i9-13900K | llama 7B Q4_0    |               3.56 |        16 | pp128  |        38.53 |                 41.96 |      1.09 |
| i9-13900K | llama 7B all F32 |              25.10 |         1 | pp32   |        30.91 |                 49.17 |      1.59 |
| i9-13900K | llama 7B all F32 |              25.10 |         1 | pp64   |        43.69 |                 61.69 |      1.41 |
| i9-13900K | llama 7B all F32 |              25.10 |         1 | pp128  |        56.80 |                 66.45 |      1.17 |
| i9-13900K | llama 7B all F32 |              25.10 |         4 | pp32   |        29.73 |                 51.31 |      1.73 |
| i9-13900K | llama 7B all F32 |              25.10 |         4 | pp64   |        44.25 |                 64.98 |      1.47 |
| i9-13900K | llama 7B all F32 |              25.10 |         4 | pp128  |        58.12 |                 74.17 |      1.28 |
| i9-13900K | llama 7B all F32 |              25.10 |         8 | pp32   |        28.94 |                 49.29 |      1.70 |
| i9-13900K | llama 7B all F32 |              25.10 |         8 | pp64   |        45.10 |                 68.30 |      1.51 |
| i9-13900K | llama 7B all F32 |              25.10 |         8 | pp128  |        58.40 |                 78.17 |      1.34 |
| i9-13900K | llama 7B all F32 |              25.10 |        16 | pp32   |        29.68 |                 32.29 |      1.09 |
| i9-13900K | llama 7B all F32 |              25.10 |        16 | pp64   |        43.85 |                 44.14 |      1.01 |
| i9-13900K | llama 7B all F32 |              25.10 |        16 | pp128  |        59.05 |                 58.07 |      0.98 |


OpenBLAS is a headache. Increasing the number of threads above 8 has catastrophic effects on performance in ways that I cannot explain, because these threads are not running while the OpenBLAS gemm is running. Since there are better alternatives easily available now, I would suggest ignoring this library.

`LLAMA_OPENBLAS=1 OPENBLAS_NUM_THREADS=24`:

| CPU       | Model            |   Model Size [GiB] |   Threads | Test   |   t/s master |   t/s sl/blas-backend |   Speedup |
|:----------|:-----------------|-------------------:|----------:|:-------|-------------:|----------------------:|----------:|
| i9-13900K | llama 7B Q4_0    |               3.56 |         4 | pp32   |        14.79 |                 14.63 |      0.99 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         4 | pp64   |        22.30 |                 21.85 |      0.98 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         4 | pp128  |        29.36 |                 28.44 |      0.97 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         8 | pp32   |        15.01 |                 15.83 |      1.05 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         8 | pp64   |        22.22 |                 23.77 |      1.07 |
| i9-13900K | llama 7B Q4_0    |               3.56 |         8 | pp128  |        29.38 |                 29.97 |      1.02 |
| i9-13900K | llama 7B all F32 |              25.10 |         4 | pp32   |        23.55 |                 23.55 |      1.00 |
| i9-13900K | llama 7B all F32 |              25.10 |         4 | pp64   |        30.88 |                 30.46 |      0.99 |
| i9-13900K | llama 7B all F32 |              25.10 |         4 | pp128  |        36.20 |                 34.92 |      0.96 |
| i9-13900K | llama 7B all F32 |              25.10 |         8 | pp32   |        23.60 |                 22.81 |      0.97 |
| i9-13900K | llama 7B all F32 |              25.10 |         8 | pp64   |        30.84 |                 31.13 |      1.01 |
| i9-13900K | llama 7B all F32 |              25.10 |         8 | pp128  |        34.42 |                 35.92 |      1.04 |


The OpenMP thread pool does not work on macOS, so the overhead of starting threads continuously is expected to be higher. However, with the default number of threads on my system (12), it still seems to result in an speedup. There is a drop in performance specifically with 4 threads and the larger batch sizes that I don't understand, I would expect the overhead to be proportionally smaller as the batch size increases.

| CPU   | Model            |   Model Size [GiB] |   Threads | Test   |   t/s master |   t/s sl/blas-backend |   Speedup |
|:------|:-----------------|-------------------:|----------:|:-------|-------------:|----------------------:|----------:|
|       | llama 7B Q4_0    |               3.56 |         4 | pp32   |        33.75 |                 35.89 |      1.06 |
|       | llama 7B Q4_0    |               3.56 |         4 | pp64   |        57.97 |                 61.30 |      1.06 |
|       | llama 7B Q4_0    |               3.56 |         4 | pp128  |        89.04 |                 91.20 |      1.02 |
|       | llama 7B Q4_0    |               3.56 |         4 | pp256  |       121.10 |                111.30 |      0.92 |
|       | llama 7B Q4_0    |               3.56 |         4 | pp512  |       128.46 |                117.87 |      0.92 |
|       | llama 7B Q4_0    |               3.56 |        12 | pp32   |        33.56 |                 37.09 |      1.11 |
|       | llama 7B Q4_0    |               3.56 |        12 | pp64   |        57.27 |                 63.48 |      1.11 |
|       | llama 7B Q4_0    |               3.56 |        12 | pp128  |        88.55 |                 97.80 |      1.10 |
|       | llama 7B Q4_0    |               3.56 |        12 | pp256  |       121.35 |                127.76 |      1.05 |
|       | llama 7B Q4_0    |               3.56 |        12 | pp512  |       128.42 |                136.79 |      1.07 |
|       | llama 7B all F32 |              25.10 |         4 | pp32   |        51.04 |                 49.88 |      0.98 |
|       | llama 7B all F32 |              25.10 |         4 | pp64   |        79.66 |                 77.13 |      0.97 |
|       | llama 7B all F32 |              25.10 |         4 | pp128  |       103.58 |                107.29 |      1.04 |
|       | llama 7B all F32 |              25.10 |         4 | pp256  |       141.23 |                123.31 |      0.87 |
|       | llama 7B all F32 |              25.10 |         4 | pp512  |       138.78 |                122.95 |      0.89 |
|       | llama 7B all F32 |              25.10 |        12 | pp32   |        50.99 |                 48.56 |      0.95 |
|       | llama 7B all F32 |              25.10 |        12 | pp64   |        79.38 |                 77.12 |      0.97 |
|       | llama 7B all F32 |              25.10 |        12 | pp128  |       112.81 |                112.74 |      1.00 |
|       | llama 7B all F32 |              25.10 |        12 | pp256  |       141.26 |                140.00 |      0.99 |
|       | llama 7B all F32 |              25.10 |        12 | pp512  |       138.73 |                140.82 |      1.02 |

slaren committed:
Merge remote-tracking branch 'origin/master' into sl/blas-backend

slaren committed:
set number of threads automatically for openblas and blis

slaren committed:
sched : print assignments when GGML_SCHED_DEBUG env variable is set

ggerganov commented:
> There is a drop in performance specifically with 4 threads and the larger batch sizes that I don't understand

On M2 Ultra there is a similar effect with the `Q4_0` and `F32` models, while for `Q8_0` and `F16` it behaves as you expect:

```bash
LLAMA_NO_LLAMAFILE=1 LLAMA_NO_METAL=1 ./scripts/compare-commits.sh master sl/blas-backend -m models/tinyllama-1b/ggml-model-q4_0.gguf -m models/tinyllama-1b/ggml-model-q8_0.gguf -m models/tinyllama-1b/ggml-model-f16.gguf -m models/tinyllama-1b/ggml-model-f32.gguf -p 32,64,128,256,512 -n 0 -t 4,8,16
```

| CPU   | Model            |   Model [GiB] |   Threads | Test   |   t/s master |   t/s sl/blas-backend |   Speedup |
|:------|:-----------------|-------------------:|----------:|:-------|-------------:|----------------------:|----------:|
| M2 Ultra | 1B F16     |               2.05 |         4 | pp32   |       265.26 |                214.52 |      0.81 |
| M2 Ultra | 1B F16     |               2.05 |         4 | pp64   |       452.51 |                361.71 |      0.80 |
| M2 Ultra | 1B F16     |               2.05 |         4 | pp128  |       640.73 |                537.69 |      0.84 |
| M2 Ultra | 1B F16     |               2.05 |         4 | pp256  |       664.02 |                588.17 |      0.89 |
| M2 Ultra | 1B F16     |               2.05 |         4 | pp512  |       624.32 |                570.26 |      0.91 |
| M2 Ultra | 1B F16     |               2.05 |         8 | pp32   |       266.72 |                263.57 |      0.99 |
| M2 Ultra | 1B F16     |               2.05 |         8 | pp64   |       464.59 |                441.93 |      0.95 |
| M2 Ultra | 1B F16     |               2.05 |         8 | pp128  |       648.08 |                660.74 |      1.02 |
| M2 Ultra | 1B F16     |               2.05 |         8 | pp256  |       659.51 |                705.26 |      1.07 |
| M2 Ultra | 1B F16     |               2.05 |         8 | pp512  |       622.22 |                745.74 |      1.20 |
| M2 Ultra | 1B F16     |               2.05 |        16 | pp32   |       271.83 |                224.23 |      0.82 |
| M2 Ultra | 1B F16     |               2.05 |        16 | pp64   |       456.11 |                394.74 |      0.87 |
| M2 Ultra | 1B F16     |               2.05 |        16 | pp128  |       633.06 |                610.69 |      0.96 |
| M2 Ultra | 1B F16     |               2.05 |        16 | pp256  |       656.49 |                699.64 |      1.07 |
| M2 Ultra | 1B F16     |               2.05 |        16 | pp512  |       625.12 |                816.75 |      1.31 |
| M2 Ultra | 1B Q4_0    |               0.59 |         4 | pp32   |       227.12 |                224.63 |      0.99 |
| M2 Ultra | 1B Q4_0    |               0.59 |         4 | pp64   |       373.63 |                371.20 |      0.99 |
| M2 Ultra | 1B Q4_0    |               0.59 |         4 | pp128  |       622.08 |                571.99 |      0.92 |
| M2 Ultra | 1B Q4_0    |               0.59 |         4 | pp256  |       658.47 |                612.67 |      0.93 |
| M2 Ultra | 1B Q4_0    |               0.59 |         4 | pp512  |       634.58 |                587.23 |      0.93 |
| M2 Ultra | 1B Q4_0    |               0.59 |         8 | pp32   |       253.38 |                269.27 |      1.06 |
| M2 Ultra | 1B Q4_0    |               0.59 |         8 | pp64   |       410.88 |                436.95 |      1.06 |
| M2 Ultra | 1B Q4_0    |               0.59 |         8 | pp128  |       636.57 |                673.07 |      1.06 |
| M2 Ultra | 1B Q4_0    |               0.59 |         8 | pp256  |       661.17 |                712.17 |      1.08 |
| M2 Ultra | 1B Q4_0    |               0.59 |         8 | pp512  |       631.02 |                745.85 |      1.18 |
| M2 Ultra | 1B Q4_0    |               0.59 |        16 | pp32   |       246.48 |                213.06 |      0.86 |
| M2 Ultra | 1B Q4_0    |               0.59 |        16 | pp64   |       396.73 |                375.92 |      0.95 |
| M2 Ultra | 1B Q4_0    |               0.59 |        16 | pp128  |       632.46 |                600.46 |      0.95 |
| M2 Ultra | 1B Q4_0    |               0.59 |        16 | pp256  |       658.53 |                703.00 |      1.07 |
| M2 Ultra | 1B Q4_0    |               0.59 |        16 | pp512  |       631.18 |                816.15 |      1.29 |
| M2 Ultra | 1B Q8_0    |               1.09 |         4 | pp32   |       248.64 |                221.43 |      0.89 |
| M2 Ultra | 1B Q8_0    |               1.09 |         4 | pp64   |       430.98 |                378.20 |      0.88 |
| M2 Ultra | 1B Q8_0    |               1.09 |         4 | pp128  |       628.65 |                565.11 |      0.90 |
| M2 Ultra | 1B Q8_0    |               1.09 |         4 | pp256  |       650.70 |                606.51 |      0.93 |
| M2 Ultra | 1B Q8_0    |               1.09 |         4 | pp512  |       629.00 |                584.55 |      0.93 |
| M2 Ultra | 1B Q8_0    |               1.09 |         8 | pp32   |       253.84 |                253.78 |      1.00 |
| M2 Ultra | 1B Q8_0    |               1.09 |         8 | pp64   |       434.08 |                431.23 |      0.99 |
| M2 Ultra | 1B Q8_0    |               1.09 |         8 | pp128  |       628.08 |                663.30 |      1.06 |
| M2 Ultra | 1B Q8_0    |               1.09 |         8 | pp256  |       645.79 |                710.54 |      1.10 |
| M2 Ultra | 1B Q8_0    |               1.09 |         8 | pp512  |       629.27 |                753.74 |      1.20 |
| M2 Ultra | 1B Q8_0    |               1.09 |        16 | pp32   |       252.36 |                214.02 |      0.85 |
| M2 Ultra | 1B Q8_0    |               1.09 |        16 | pp64   |       431.50 |                372.52 |      0.86 |
| M2 Ultra | 1B Q8_0    |               1.09 |        16 | pp128  |       635.57 |                589.64 |      0.93 |
| M2 Ultra | 1B Q8_0    |               1.09 |        16 | pp256  |       644.00 |                707.12 |      1.10 |
| M2 Ultra | 1B Q8_0    |               1.09 |        16 | pp512  |       634.01 |                813.76 |      1.28 |
| M2 Ultra | 1B all F32 |               4.10 |         4 | pp32   |       324.34 |                318.84 |      0.98 |
| M2 Ultra | 1B all F32 |               4.10 |         4 | pp64   |       497.14 |                501.49 |      1.01 |
| M2 Ultra | 1B all F32 |               4.10 |         4 | pp128  |       780.47 |                687.67 |      0.88 |
| M2 Ultra | 1B all F32 |               4.10 |         4 | pp256  |       719.40 |                676.08 |      0.94 |
| M2 Ultra | 1B all F32 |               4.10 |         4 | pp512  |       668.30 |                614.47 |      0.92 |
| M2 Ultra | 1B all F32 |               4.10 |         8 | pp32   |       394.99 |                325.57 |      0.82 |
| M2 Ultra | 1B all F32 |               4.10 |         8 | pp64   |       588.41 |                522.05 |      0.89 |
| M2 Ultra | 1B all F32 |               4.10 |         8 | pp128  |       769.56 |                727.87 |      0.95 |
| M2 Ultra | 1B all F32 |               4.10 |         8 | pp256  |       715.89 |                742.46 |      1.04 |
| M2 Ultra | 1B all F32 |               4.10 |         8 | pp512  |       672.34 |                772.63 |      1.15 |
| M2 Ultra | 1B all F32 |               4.10 |        16 | pp32   |       397.21 |                246.50 |      0.62 |
| M2 Ultra | 1B all F32 |               4.10 |        16 | pp64   |       589.79 |                421.42 |      0.71 |
| M2 Ultra | 1B all F32 |               4.10 |        16 | pp128  |       770.60 |                655.37 |      0.85 |
| M2 Ultra | 1B all F32 |               4.10 |        16 | pp256  |       712.96 |                724.44 |      1.02 |
| M2 Ultra | 1B all F32 |               4.10 |        16 | pp512  |       669.12 |                837.94 |      1.25 |

slaren committed:
fixes

slaren commented:
I realized that there is an issue that causes the `kqv` matrix multiplication to not be offloaded to the BLAS backend, and this is what causes the performance difference with a low number of threads. If this matrix multiplication is removed by enabling flash attention, then the performance difference disappears.
| CPU   | Model            |   Model Size [GiB] |   Threads | Test   |   t/s master |   t/s sl/blas-backend |   Speedup |
|:------|:-----------------|-------------------:|----------:|:-------|-------------:|----------------------:|----------:|
| M3 Max | 7B Q4_0    |               3.56 |         4 | pp32   |        32.77 |                 36.23 |      1.11 |
| M3 Max | 7B Q4_0    |               3.56 |         4 | pp64   |        58.37 |                 62.11 |      1.06 |
| M3 Max | 7B Q4_0    |               3.56 |         4 | pp128  |        92.00 |                 95.05 |      1.03 |
| M3 Max | 7B Q4_0    |               3.56 |         4 | pp256  |       119.94 |                121.50 |      1.01 |
| M3 Max | 7B Q4_0    |               3.56 |         4 | pp512  |       127.86 |                126.98 |      0.99 |
| M3 Max | 7B Q4_0    |               3.56 |        12 | pp32   |        33.67 |                 38.43 |      1.14 |
| M3 Max | 7B Q4_0    |               3.56 |        12 | pp64   |        57.92 |                 64.87 |      1.12 |
| M3 Max | 7B Q4_0    |               3.56 |        12 | pp128  |        90.62 |                 99.64 |      1.10 |
| M3 Max | 7B Q4_0    |               3.56 |        12 | pp256  |       120.33 |                129.79 |      1.08 |
| M3 Max | 7B Q4_0    |               3.56 |        12 | pp512  |       126.31 |                141.03 |      1.12 |
| M3 Max | 7B F32 |              25.10 |         4 | pp32   |        51.18 |                 50.64 |      0.99 |
| M3 Max | 7B F32 |              25.10 |         4 | pp64   |        79.80 |                 79.86 |      1.00 |
| M3 Max | 7B F32 |              25.10 |         4 | pp128  |       112.13 |                111.49 |      0.99 |
| M3 Max | 7B F32 |              25.10 |         4 | pp256  |       138.19 |                133.77 |      0.97 |
| M3 Max | 7B F32 |              25.10 |         4 | pp512  |       133.84 |                131.51 |      0.98 |
| M3 Max | 7B F32 |              25.10 |        12 | pp32   |        50.47 |                 48.97 |      0.97 |
| M3 Max | 7B F32 |              25.10 |        12 | pp64   |        79.65 |                 78.19 |      0.98 |
| M3 Max | 7B F32 |              25.10 |        12 | pp128  |       112.97 |                115.07 |      1.02 |
| M3 Max | 7B F32 |              25.10 |        12 | pp256  |       137.53 |                143.82 |      1.05 |
| M3 Max | 7B F32 |              25.10 |        12 | pp512  |       134.25 |                149.98 |      1.12 |

slaren commented:
This should be good now. I have updated the PR description with more details about the changes included here.

ggerganov reviewed (commented):
> Note: the BLAS backend should not be used alongside GPU backends, as it will prevent offloading of large batches with partial offloading

On macOS with Metal enabled, when I build with `LLAMA_BLAS=OFF` and run with partial offloading (`-ngl 28`), the non-offloaded layers are running on the CPU backend:

```
...
node # 32 (       ADD):              l_out-0 (   8M) [  CPU         ]:            ffn_out-0 (   8M) [  CPU         ]            ffn_inp-0 (   8M) [  CPU         ]
node # 33 (  RMS_NORM):               norm-1 (   8M) [  CPU         ]:              l_out-0 (   8M) [  CPU         ]
node # 34 (       MUL):          attn_norm-1 (   8M) [  CPU         ]:               norm-1 (   8M) [  CPU         ] blk.1.attn_norm.weig (  16K) [  CPU         ]
node # 35 (   MUL_MAT):               Qcur-1 (   8M) [  CPU         ]:  blk.1.attn_q.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [  CPU         ]
node # 37 (      ROPE):               Qcur-1 (   8M) [  CPU         ]:    Qcur-1 (reshaped) (   8M) [  CPU         ]              inp_pos (   2K) [  CPU         ]
node # 38 (   MUL_MAT):               Kcur-1 (   8M) [  CPU         ]:  blk.1.attn_k.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [  CPU         ]
node # 40 (      ROPE):               Kcur-1 (   8M) [  CPU         ]:    Kcur-1 (reshaped) (   8M) [  CPU         ]              inp_pos (   2K) [  CPU         ]
node # 41 (   MUL_MAT):               Vcur-1 (   8M) [  CPU         ]:  blk.1.attn_v.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [  CPU         ]
node # 43 (       CPY): k_cache_view-1 (copy (   4M) [  CPU         ]:               Kcur-1 (   8M) [  CPU         ]       k_cache_view-1 (   4M) [  CPU         ]
node # 46 (       CPY): v_cache_view-1 (copy (   4M) [  CPU         ]:  Vcur-1 (transposed) (   8M) [  CPU         ]       v_cache_view-1 (   4M) [  CPU         ]
node # 50 (   MUL_MAT):                 kq-1 (  32M) [  CPU         ]:                  k-1 (   4M) [  CPU         ]                  q-1 (   8M) [  CPU         ]
node # 51 (  SOFT_MAX):    kq_soft_max_ext-1 (  32M) [  CPU         ]:                 kq-1 (  32M) [  CPU         ]              KQ_mask (   1M) [  CPU         ]
node # 52 (   MUL_MAT):                kqv-1 (   8M) [  CPU         ]:                  v-1 (   4M) [  CPU         ]    kq_soft_max_ext-1 (  32M) [  CPU         ]
node # 54 (      CONT):    kqv_merged_cont-1 (   8M) [  CPU         ]:         kqv_merged-1 (   8M) [  CPU         ]
node # 55 (   MUL_MAT):            kqv_out-1 (   8M) [  CPU         ]: blk.1.attn_output.we (  17M) [  CPU         ]    kqv_merged_cont-1 (   8M) [  CPU         ]
node # 56 (       ADD):            ffn_inp-1 (   8M) [  CPU         ]:            kqv_out-1 (   8M) [  CPU         ]              l_out-0 (   8M) [  CPU         ]
node # 57 (  RMS_NORM):               norm-1 (   8M) [  CPU         ]:            ffn_inp-1 (   8M) [  CPU         ]
node # 58 (       MUL):           ffn_norm-1 (   8M) [  CPU         ]:               norm-1 (   8M) [  CPU         ] blk.1.ffn_norm.weigh (  16K) [  CPU         ]
node # 59 (   MUL_MAT):           ffn_gate-1 (  21M) [  CPU         ]: blk.1.ffn_gate.weigh (  45M) [  CPU         ]           ffn_norm-1 (   8M) [  CPU         ]
node # 60 (     UNARY):           ffn_silu-1 (  21M) [  CPU         ]:           ffn_gate-1 (  21M) [  CPU         ]
node # 61 (   MUL_MAT):             ffn_up-1 (  21M) [  CPU         ]:  blk.1.ffn_up.weight (  45M) [  CPU         ]           ffn_norm-1 (   8M) [  CPU         ]
node # 62 (       MUL):       ffn_gate_par-1 (  21M) [  CPU         ]:           ffn_silu-1 (  21M) [  CPU         ]             ffn_up-1 (  21M) [  CPU         ]
node # 63 (   MUL_MAT):            ffn_out-1 (   8M) [  CPU         ]: blk.1.ffn_down.weigh (  45M) [  CPU         ]       ffn_gate_par-1 (  21M) [  CPU         ]
node # 64 (       ADD):              l_out-1 (   8M) [  CPU         ]:            ffn_out-1 (   8M) [  CPU         ]            ffn_inp-1 (   8M) [  CPU         ]
node # 65 (  RMS_NORM):               norm-2 (   8M) [  CPU         ]:              l_out-1 (   8M) [  CPU         ]
...
```

With `LLAMA_BLAS=ON` it uses the BLAS backend for the matrix multiplications:

```
...
## SPLIT #16: Metal # 1 inputs: [ffn_out-0 (   8M)] 
node # 32 (       ADD):              l_out-0 (   8M) [Metal         ]:    Metal#ffn_out-0#0 (   8M) [ NULL         ]            ffn_inp-0 (   8M) [Metal         ]
node # 33 (  RMS_NORM):               norm-1 (   8M) [Metal         ]:              l_out-0 (   8M) [Metal         ]

## SPLIT #17: CPU # 0 inputs: 
node # 34 (       MUL):          attn_norm-1 (   8M) [  CPU         ]:               norm-1 (   8M) [Metal         ] blk.1.attn_norm.weig (  16K) [  CPU         ]

## SPLIT #18: BLAS # 0 inputs: 
node # 35 (   MUL_MAT):               Qcur-1 (   8M) [ BLAS         ]:  blk.1.attn_q.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [  CPU         ]

## SPLIT #19: Metal # 1 inputs: [Qcur-1 (reshaped) (   8M)] 
node # 37 (      ROPE):               Qcur-1 (   8M) [Metal         ]: Metal#Qcur-1 (reshap (   8M) [ NULL         ]      Metal#inp_pos#0 (   2K) [ NULL         ]

## SPLIT #20: BLAS # 0 inputs: 
node # 38 (   MUL_MAT):               Kcur-1 (   8M) [ BLAS         ]:  blk.1.attn_k.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [  CPU         ]

## SPLIT #21: Metal # 1 inputs: [Kcur-1 (reshaped) (   8M)] 
node # 40 (      ROPE):               Kcur-1 (   8M) [Metal         ]: Metal#Kcur-1 (reshap (   8M) [ NULL         ]      Metal#inp_pos#0 (   2K) [ NULL         ]

## SPLIT #22: BLAS # 0 inputs: 
node # 41 (   MUL_MAT):               Vcur-1 (   8M) [ BLAS         ]:  blk.1.attn_v.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [  CPU         ]

## SPLIT #23: CPU # 0 inputs: 
node # 43 (       CPY): k_cache_view-1 (copy (   4M) [  CPU         ]:               Kcur-1 (   8M) [Metal         ]       k_cache_view-1 (   4M) [  CPU         ]
node # 46 (       CPY): v_cache_view-1 (copy (   4M) [  CPU         ]:  Vcur-1 (transposed) (   8M) [ BLAS         ]       v_cache_view-1 (   4M) [  CPU         ]

## SPLIT #24: Metal # 2 inputs: [k-1 (   4M)] [v-1 (   4M)] 
node # 50 (   MUL_MAT):                 kq-1 (  32M) [Metal         ]:          Metal#k-1#0 (   4M) [ NULL         ]                  q-1 (   8M) [Metal         ]
node # 51 (  SOFT_MAX):    kq_soft_max_ext-1 (  32M) [Metal         ]:                 kq-1 (  32M) [Metal         ]      Metal#KQ_mask#0 (   1M) [ NULL         ]
node # 52 (   MUL_MAT):                kqv-1 (   8M) [Metal         ]:          Metal#v-1#0 (   4M) [ NULL         ]    kq_soft_max_ext-1 (  32M) [Metal         ]
node # 54 (      CONT):    kqv_merged_cont-1 (   8M) [Metal         ]:         kqv_merged-1 (   8M) [Metal         ]

## SPLIT #25: BLAS # 0 inputs: 
node # 55 (   MUL_MAT):            kqv_out-1 (   8M) [ BLAS         ]: blk.1.attn_output.we (  17M) [  CPU         ]    kqv_merged_cont-1 (   8M) [Metal         ]

## SPLIT #26: Metal # 1 inputs: [kqv_out-1 (   8M)] 
node # 56 (       ADD):            ffn_inp-1 (   8M) [Metal         ]:    Metal#kqv_out-1#0 (   8M) [ NULL         ]              l_out-0 (   8M) [Metal         ]
node # 57 (  RMS_NORM):               norm-1 (   8M) [Metal         ]:            ffn_inp-1 (   8M) [Metal         ]

## SPLIT #27: CPU # 0 inputs: 
node # 58 (       MUL):           ffn_norm-1 (   8M) [  CPU         ]:               norm-1 (   8M) [Metal         ] blk.1.ffn_norm.weigh (  16K) [  CPU         ]

## SPLIT #28: BLAS # 0 inputs: 
node # 59 (   MUL_MAT):           ffn_gate-1 (  21M) [ BLAS         ]: blk.1.ffn_gate.weigh (  45M) [  CPU         ]           ffn_norm-1 (   8M) [  CPU         ]

## SPLIT #29: Metal # 1 inputs: [ffn_gate-1 (  21M)] 
node # 60 (     UNARY):           ffn_silu-1 (  21M) [Metal         ]:   Metal#ffn_gate-1#0 (  21M) [ NULL         ]

## SPLIT #30: BLAS # 0 inputs: 
node # 61 (   MUL_MAT):             ffn_up-1 (  21M) [ BLAS         ]:  blk.1.ffn_up.weight (  45M) [  CPU         ]           ffn_norm-1 (   8M) [  CPU         ]

## SPLIT #31: Metal # 1 inputs: [ffn_up-1 (  21M)] 
node # 62 (       MUL):       ffn_gate_par-1 (  21M) [Metal         ]:           ffn_silu-1 (  21M) [Metal         ]     Metal#ffn_up-1#0 (  21M) [ NULL         ]

## SPLIT #32: BLAS # 0 inputs: 
node # 63 (   MUL_MAT):            ffn_out-1 (   8M) [ BLAS         ]: blk.1.ffn_down.weigh (  45M) [  CPU         ]       ffn_gate_par-1 (  21M) [Metal         ]

## SPLIT #33: Metal # 1 inputs: [ffn_out-1 (   8M)] 
node # 64 (       ADD):              l_out-1 (   8M) [Metal         ]:    Metal#ffn_out-1#0 (   8M) [ NULL         ]            ffn_inp-1 (   8M) [Metal         ]
node # 65 (  RMS_NORM):               norm-2 (   8M) [Metal         ]:              l_out-1 (   8M) [Metal         ]
...
```

Is this the expectation? It seems like using BLAS together with GPU offloading leads to improvement in this case, or did I misunderstood this comment?

slaren commented:
Specifically, this applies to backends that implement the `offload_op` function to offload large batches even when the model is not completely offloaded, which currently it is CUDA, Vulkan and SYCL. For these backends, enabling the BLAS backend will cause it to be used instead of offloading large batches to the GPU by copying the weights to VRAM as needed. Since the Metal backend does not implement this function it is not affected, and the BLAS backend can be used to enable Accelerate for the layers not offloaded.

slaren committed:
Apply suggestions from code review

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

slaren commented:
Metal should not be used for the operations in between the BLAS backend in not offloaded layers though, I will try to fix that.

zhouwg commented:
> Metal should not be used for the operations in between the BLAS backend in not offloaded layers though, I will try to fix that.

pls consider my standalone PR for purpose of mixed inference between CPU&GPU / CPU&NPU if ggml_backend_xx_buffer_is_host.

zhouwg commented:
> Metal should not be used for the operations in between the BLAS backend in not offloaded layers though, I will try to fix that.

pls consider my standalone PR for purpose of mixed inference between CPU&GPU / CPU&NPU if ggml_backend_xx_buffer_is_host.

slaren commented:
@zhouwg I already considered it and rejected it. Spamming more about it is not going to help your cause.

zhouwg commented:
> Metal should not be used for the operations in between the BLAS backend in not offloaded layers though, I will try to fix that.

pls consider my [standalone PR](https://github.com/ggerganov/llama.cpp/pull/7641) for purpose of mixed inference between CPU&GPU / CPU if ggml_backend_xxx_buffer_is_host return true

> @zhouwg I already considered it and rejected it. Spamming more about it is not going to help your cause.

it's caused by the GFW. you(programmers/developers/AI experts in US/EU) can not image that.

zhouwg commented:
> @zhouwg Please focus on your PR and respect the comments and suggestions that have already been provided. Consider this final warning, before having to block you

thanks for your reminder. I see.

ggerganov commented:
In that same example, if I allow the `GGML_OP_MUL` operation to be offloaded in the Metal backend, I get the following schedule:

```
## SPLIT #7: Metal # 1 inputs: [ffn_out-0 (   8M)] 
node # 32 (       ADD):              l_out-0 (   8M) [Metal         ]:    Metal#ffn_out-0#0 (   8M) [ NULL         ]            ffn_inp-0 (   8M) [Metal         ]
node # 33 (  RMS_NORM):               norm-1 (   8M) [Metal         ]:              l_out-0 (   8M) [Metal         ]

## SPLIT #8: Metal # 1 inputs: [blk.1.attn_norm.weight (  16K)] 
node # 34 (       MUL):          attn_norm-1 (   8M) [Metal         ]:               norm-1 (   8M) [Metal         ] Metal#blk.1.attn_nor (  16K) [ NULL         ]

## SPLIT #9: CPU # 0 inputs: 
node # 35 (   MUL_MAT):               Qcur-1 (   8M) [  CPU         ]:  blk.1.attn_q.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [Metal         ]
node # 37 (      ROPE):               Qcur-1 (   8M) [  CPU         ]:    Qcur-1 (reshaped) (   8M) [  CPU         ]              inp_pos (   2K) [  CPU         ]
node # 38 (   MUL_MAT):               Kcur-1 (   8M) [  CPU         ]:  blk.1.attn_k.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [Metal         ]
node # 40 (      ROPE):               Kcur-1 (   8M) [  CPU         ]:    Kcur-1 (reshaped) (   8M) [  CPU         ]              inp_pos (   2K) [  CPU         ]
node # 41 (   MUL_MAT):               Vcur-1 (   8M) [  CPU         ]:  blk.1.attn_v.weight (  17M) [  CPU         ]          attn_norm-1 (   8M) [Metal         ]
node # 43 (       CPY): k_cache_view-1 (copy (   4M) [  CPU         ]:               Kcur-1 (   8M) [  CPU         ]       k_cache_view-1 (   4M) [  CPU         ]
node # 46 (       CPY): v_cache_view-1 (copy (   4M) [  CPU         ]:  Vcur-1 (transposed) (   8M) [  CPU         ]       v_cache_view-1 (   4M) [  CPU         ]
node # 50 (   MUL_MAT):                 kq-1 (  32M) [  CPU         ]:                  k-1 (   4M) [  CPU         ]                  q-1 (   8M) [  CPU         ]
node # 51 (  SOFT_MAX):    kq_soft_max_ext-1 (  32M) [  CPU         ]:                 kq-1 (  32M) [  CPU         ]              KQ_mask (   1M) [  CPU         ]
node # 52 (   MUL_MAT):                kqv-1 (   8M) [  CPU         ]:                  v-1 (   4M) [  CPU         ]    kq_soft_max_ext-1 (  32M) [  CPU         ]
node # 54 (      CONT):    kqv_merged_cont-1 (   8M) [  CPU         ]:         kqv_merged-1 (   8M) [  CPU         ]
node # 55 (   MUL_MAT):            kqv_out-1 (   8M) [  CPU         ]: blk.1.attn_output.we (  17M) [  CPU         ]    kqv_merged_cont-1 (   8M) [  CPU         ]

## SPLIT #10: Metal # 1 inputs: [kqv_out-1 (   8M)] 
node # 56 (       ADD):            ffn_inp-1 (   8M) [Metal         ]:    Metal#kqv_out-1#0 (   8M) [ NULL         ]              l_out-0 (   8M) [Metal         ]
node # 57 (  RMS_NORM):               norm-1 (   8M) [Metal         ]:            ffn_inp-1 (   8M) [Metal         ]

## SPLIT #11: Metal # 1 inputs: [blk.1.ffn_norm.weight (  16K)] 
node # 58 (       MUL):           ffn_norm-1 (   8M) [Metal         ]:               norm-1 (   8M) [Metal         ] Metal#blk.1.ffn_norm (  16K) [ NULL         ]

## SPLIT #12: CPU # 0 inputs: 
node # 59 (   MUL_MAT):           ffn_gate-1 (  21M) [  CPU         ]: blk.1.ffn_gate.weigh (  45M) [  CPU         ]           ffn_norm-1 (   8M) [Metal         ]
```

How does the logic decide to also offload nodes `#56 (ADD) and #57 (RMS_NORM)` in addition to `#58 (MUL)`?

slaren commented:
In the first pass, ops with weights are assigned the backend of the weight. `offload_op` is used at this point to allow overriding this assignment when the batch size is large enough that it may be worth to copy the weight to VRAM. Then these initial assignments are expanded to the rest of the ops. In this case, what is likely happening is that #58 is assigned to Metal due to `offload_op`, and then this assignment was expanded to the adjacent ops. You can enable the `GET_CAUSE/SET_CAUSE` macros to find exactly at which step the assignment was made.

slaren committed:
fix metal being used in layers not offloaded

slaren committed:
sched : allow ops with weights on an incompatible buffer type

This will cause the weight to be copied to a backend that supports the
op, which is very costly. The weight should have been stored in a buffer
of a backend that can run the op, but llama.cpp cannot do this
automatically at the moment.

ggml-ci
